airavata-api/airavata-api-server/src/main/java/org/apache/airavata/api/server/handler/AiravataServerHandler.java

```diff
@@ -1747,16 +1761,24 @@ public ExperimentStatus getExperimentStatus(AuthzToken authzToken, String airava
     @Override
     @SecurityCheck
     public void launchExperiment(AuthzToken authzToken, final String airavataExperimentId, String gatewayId)
-            throws TException {
+            throws AuthorizationException, AiravataSystemException, TException {
         RegistryService.Client regClient = registryClientPool.getResource();
+        SharingRegistryService.Client sharingClient = sharingClientPool.getResource();
         try {
             ExperimentModel experiment = regClient.getExperiment(airavataExperimentId);
+            String userId = authzToken.getClaimsMap().get(Constants.USER_NAME);
+            String entityId = experiment.getExecutionId();
+            if (!sharingClient.userHasAccess(gatewayId, userId + "@" + gatewayId, entityId,gatewayId + ":READ")) {
+                logger.error(airavataExperimentId, "User does not have access to application module {}.", entityId);
+                throw new AuthorizationException("User does not have permission to access this resource");
+            }
             if (experiment == null) {
                 logger.error(airavataExperimentId, "Error while launching experiment, experiment {} doesn't exist.", airavataExperimentId);
                 throw new ExperimentNotFoundException("Requested experiment id " + airavataExperimentId + " does not exist in the system..");
             }
             submitExperiment(gatewayId, airavataExperimentId);
             registryClientPool.returnResource(regClient);
+            sharingClientPool.returnResource(sharingClient);
```

Comment:
You should probably return the sharingClient to the pool as a broken resource, like what is done for the regClient, see line 1787.

Reply:
Done.